fix!: Fix handling the CMEK with automatic replication

legal90 opened the pull request:
Fixes https://github.com/GoogleCloudPlatform/terraform-google-secret-manager/issues/73

## Changes
- Fixed the conditions for `auto / customer_managed_encryption` dynamic blocks, so the value of `kms_key_name` is properly passed to the resource `google_secret_manager_secret `
- BREAKING CHANGE: Changed the type of `automatic_replication` variable, removing the redundant `list()` nesting level. There could be only one `auto {...}` block in `replication`,
- Removed `automatic_replication = {}` from example, since that it a default value and it could be omitted,

## Example

```hcl
# main.tf

module "secret-manager" {
  source = "../../"

  project_id = var.project_id
  secrets = [{
    name        = "secret-1"
    secret_data = "secret information"
  }]
  automatic_replication = {
    secret-1 = {     # <= Now the type is `map(object(...))` instead of `map(list(object(...)))`
      kms_key_name = "projects/my-project/locations/global/keyRings/key-ring-1/cryptoKeys/crypto-key-1"
    }
  }
}
```

### Result

```shell
$ terraform plan

  # module.secret-manager.google_secret_manager_secret.secrets["secret-1"] will be created
  + resource "google_secret_manager_secret" "secrets" {
      + create_time           = (known after apply)
      + effective_annotations = (known after apply)
      + effective_labels      = (known after apply)
      + expire_time           = (known after apply)
      + id                    = (known after apply)
      + name                  = (known after apply)
      + project               = "<my-project>"
      + secret_id             = "secret-1"
      + terraform_labels      = (known after apply)

      + replication {
          + auto {
              + customer_managed_encryption {
                  + kms_key_name = "projects/my-project/locations/global/keyRings/key-ring-1/cryptoKeys/crypto-key-1"
                }
            }
        }
    }
```

apeabody commented:
/gcbrun

apeabody commented:
From int test:

```
Error: Error creating Secret: googleapi: Error 400: Secret Manager service identity for project not found. A service identity must be created to use customer-managed encryption keys. Please see https://cloud.google.com/secret-manager/docs/cmek for details on how to create a service identity.

  with module.secret-manager.google_secret_manager_secret.secrets["secret-multi-1"],
  on ../../main.tf line 39, in resource "google_secret_manager_secret" "secrets":
  39: resource "google_secret_manager_secret" "secrets" {


Error: Error creating Secret: googleapi: Error 400: Automatic replication policy can only be configured with Cloud KMS keys in location [global]. Requested: [projects/ci-secret-manager-ea10/locations/us-east1/keyRings/key-ring-east-c989/cryptoKeys/crypto-key-c989]

  with module.secret-manager.google_secret_manager_secret.secrets["secret-2"],
  on ../../main.tf line 39, in resource "google_secret_manager_secret" "secrets":
  39: resource "google_secret_manager_secret" "secrets" {
}
    apply.go:34: 
        	Error Trace:	/builder/home/go/pkg/mod/github.com/gruntwork-io/terratest@v0.46.15/modules/terraform/apply.go:34
        	            				/builder/home/go/pkg/mod/github.com/!google!cloud!platform/cloud-foundation-toolkit/infra/blueprint-test@v0.15.1/pkg/tft/terraform.go:538
        	            				/builder/home/go/pkg/mod/github.com/!google!cloud!platform/cloud-foundation-toolkit/infra/blueprint-test@v0.15.1/pkg/tft/terraform.go:597
        	            				/builder/home/go/pkg/mod/github.com/!google!cloud!platform/cloud-foundation-toolkit/infra/blueprint-test@v0.15.1/pkg/tft/terraform.go:636
        	            				/builder/home/go/pkg/mod/github.com/!google!cloud!platform/cloud-foundation-toolkit/infra/blueprint-test@v0.15.1/pkg/utils/stages.go:31
        	            				/builder/home/go/pkg/mod/github.com/!google!cloud!platform/cloud-foundation-toolkit/infra/blueprint-test@v0.15.1/pkg/tft/terraform.go:636
        	Error:      	Received unexpected error:
        	            	FatalError{Underlying: error while running command: exit status 1; 
        	            	Error: Error creating Secret: googleapi: Error 400: Secret Manager service identity for project not found. A service identity must be created to use customer-managed encryption keys. Please see https://cloud.google.com/secret-manager/docs/cmek for details on how to create a service identity.
        	            	
        	            	  with module.secret-manager.google_secret_manager_secret.secrets["secret-multi-1"],
        	            	  on ../../main.tf line 39, in resource "google_secret_manager_secret" "secrets":
        	            	  39: resource "google_secret_manager_secret" "secrets" {
        	            	
        	            	
        	            	Error: Error creating Secret: googleapi: Error 400: Automatic replication policy can only be configured with Cloud KMS keys in location [global]. Requested: [projects/ci-secret-manager-ea10/locations/us-east1/keyRings/key-ring-east-c989/cryptoKeys/crypto-key-c989]
        	            	
        	            	  with module.secret-manager.google_secret_manager_secret.secrets["secret-2"],
        	            	  on ../../main.tf line 39, in resource "google_secret_manager_secret" "secrets":
        	            	  39: resource "google_secret_manager_secret" "secrets" {
        	            	}
        	Test:       	TestAll/examples/multiple
```

legal90 commented:
Thank you, @apeabody !

> Error: Error creating Secret: googleapi: Error 400: Automatic replication policy can only be configured with Cloud KMS keys in location [global]. Requested: [projects/ci-secret-manager-ea10/locations/us-east1/keyRings/key-ring-east-c989/cryptoKeys/crypto-key-c989]

^ I fixed that error now in https://github.com/GoogleCloudPlatform/terraform-google-secret-manager/pull/74/commits/ec3d746d7f600df2b2c9b2a1ad55fb1aea7f3575

> Error: Error creating Secret: googleapi: Error 400: Secret Manager service identity for project not found. A service identity must be created to use customer-managed encryption keys. Please see https://cloud.google.com/secret-manager/docs/cmek for details on how to create a service identity.

 ^ It seems that we have to add [project_service_identity](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/project_service_identity) resource when we want to use CMEK for Secret encryption. 
What do you think - should I add it to the root module (and thus it will always be created for all users, regardless of the configuration)? Or should I just add it to that particular example test "multiple" (and users will need to do the same on their side only when they want to use CMEK)?

apeabody commented:
/gcbrun

apeabody commented:
> Thank you, @apeabody !
> 
> > Error: Error creating Secret: googleapi: Error 400: Automatic replication policy can only be configured with Cloud KMS keys in location [global]. Requested: [projects/ci-secret-manager-ea10/locations/us-east1/keyRings/key-ring-east-c989/cryptoKeys/crypto-key-c989]
> 
> ^ I fixed that error now in [ec3d746](https://github.com/GoogleCloudPlatform/terraform-google-secret-manager/commit/ec3d746d7f600df2b2c9b2a1ad55fb1aea7f3575)
> 
> > Error: Error creating Secret: googleapi: Error 400: Secret Manager service identity for project not found. A service identity must be created to use customer-managed encryption keys. Please see https://cloud.google.com/secret-manager/docs/cmek for details on how to create a service identity.
> 
> ^ It seems that we have to add [project_service_identity](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/project_service_identity) resource when we want to use CMEK for Secret encryption. What do you think - should I add it to the root module (and thus it will always be created for all users, regardless of the configuration)? Or should I just add it to that particular example test "multiple" (and users will need to do the same on their side only when they want to use CMEK)?

Thanks @legal90!

My initial thought would be the later, adding to the example, as that would minimize the size/impact of the change and also avoid any potential issues with duplication of the project_service_identity resource.

apeabody commented:
```
        	Error:      	Received unexpected error:
        	            	FatalError{Underlying: error while running command: exit status 1; 
        	            	Error: Error creating Secret: googleapi: Error 400: Permission denied on Cloud KMS resource [projects/ci-secret-manager-e7a5/locations/global/keyRings/key-ring-east-bbbf/cryptoKeys/crypto-key-bbbf] (or it does not exist). Please grant cloudkms.cryptoKeyVersions.useToDecrypt and cloudkms.cryptoKeyVersions.useToEncrypt permissions (roles/cloudkms.cryptoKeyEncrypterDecrypter) to the Secret Manager service identity. See https://cloud.google.com/secret-manager/docs/cmek for more information.
        	            	
        	            	  with module.secret-manager.google_secret_manager_secret.secrets["secret-2"],
        	            	  on ../../main.tf line 39, in resource "google_secret_manager_secret" "secrets":
        	            	  39: resource "google_secret_manager_secret" "secrets" {
        	            	}
        	Test:       	TestAll/examples/multiple
```

legal90 commented:
@apeabody I just realized that the root module already has `project_service_identity` resource, so it is already created and should work as expected.
The only missing part was the KMS key permission in that test example. I fixed it in https://github.com/GoogleCloudPlatform/terraform-google-secret-manager/pull/74/commits/9bcf3cb03d7599b9d91502180cefe03876e429b3
Could you please check is the test passes now?

apeabody commented:
/gcbrun

apeabody reviewed (approved):
Thanks for the contribution @legal90!

We should probably mark this as a breaking change given the change the variable syntax?

legal90 commented:
@apeabody OK, I changed the title `fix:` -> `fix!:`  and added _BREAKING CHANGE:_ disclaimer to the first message.
Should I also reword commit message(s)?

apeabody commented:
> @apeabody OK, I changed the PR title `fix:` -> `fix!:` and added _BREAKING CHANGE:_ disclaimer to the first message. Should I also reword commit message(s)?

Thanks! - No need to change the commit, we squash to the PR title anyway.
